lib/vrl/stdlib/src/parse_ruby_hash.rs

```diff
@@ -0,0 +1,264 @@
+        let value = arguments.required("value");
+        Ok(Box::new(ParseRubyHashFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::BYTES,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseRubyHashFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ParseRubyHashFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let input = value.try_bytes_utf8_lossy()?;
+        parse(&input)
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new().infallible().object::<(), Kind>(type_def())
+    }
+}
+
+fn type_def() -> BTreeMap<(), Kind> {
+    map! {
+        (): Kind::Bytes | Kind::Array,
+    }
+}
+
+fn sp<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, &'a str, E> {
+    let chars = " \t\r\n";
+
+    take_while(move |c| chars.contains(c))(input)
+}
+
+fn parse_str<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, &'a str, E> {
+    escaped(alphanumeric, '\\', one_of("\"n\\"))(input)
+}
+
+fn parse_boolean<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, bool, E> {
+    let parse_true = value(true, tag("true"));
+    let parse_false = value(false, tag("false"));
+
+    alt((parse_true, parse_false))(input)
+}
+
+fn parse_nil<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, Value, E> {
+    value(Value::Null, tag("nil"))(input)
+}
+
+fn parse_bytes<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
+    input: &'a str,
+) -> IResult<&'a str, Bytes, E> {
+    context(
+        "bytes",
+        map(
+            preceded(char('\"'), cut(terminated(parse_str::<'a, E>, char('\"')))),
+            |value| Bytes::copy_from_slice(value.as_bytes()),
+        ),
+    )(input)
+}
+
+fn parse_key<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
+    input: &'a str,
+) -> IResult<&'a str, String, E> {
+    context(
+        "string",
+        map(
+            alt((
+                preceded(char('\"'), cut(terminated(parse_str, char('\"')))),
+                preceded(char('\''), cut(terminated(parse_str, char('\'')))),
+                alphanumeric,
+            )),
+            String::from,
+        ),
+    )(input)
+}
+
+fn parse_array<
+    'a,
+    E: ParseError<&'a str> + ContextError<&'a str> + FromExternalError<&'a str, ParseIntError>,
+>(
+    input: &'a str,
+) -> IResult<&'a str, Vec<Value>, E> {
+    context(
+        "array",
+        preceded(
+            char('['),
+            cut(terminated(
+                separated_list0(preceded(sp, char(',')), parse_value),
+                preceded(sp, char(']')),
+            )),
+        ),
+    )(input)
+}
+
+fn parse_key_value<
+    'a,
+    E: ParseError<&'a str> + ContextError<&'a str> + FromExternalError<&'a str, ParseIntError>,
```

Comment:
Might be good to move `ParseError<&'a str> + ContextError<&'a str> + FromExternalError<&'a str, ParseIntError>` into a type alias, to clean up the code a bit.

Reply:
you mean a `trait` alias?

Reply:
Ah yeah, Trait aliases [aren't a thing yet](https://doc.rust-lang.org/beta/unstable-book/language-features/trait-alias.html).

You can create a new trait to capture all the relevant trait requirements though: https://stackoverflow.com/a/26071172

Reply:
Thank you, I was missing the `impl<T: PartialOrd + Display> PartialDisplay for T {}`